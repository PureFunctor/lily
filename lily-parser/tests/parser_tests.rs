use lily_parser::parse_top_level;

#[test]
pub fn top_level_0() {
    let source = r"
infixl 4 add as +

example = a + b + c
";

    insta::assert_debug_snapshot!(parse_top_level(source));
}

#[test]
pub fn top_level_1() {
    let source = r"
example = a + b + c
";

    insta::assert_debug_snapshot!(parse_top_level(source));
}

#[test]
pub fn top_level_2() {
    let source = r"
infixl 1 add as +
infixl 2 add as *

example = a + b * (c + d) + e
";

    insta::assert_debug_snapshot!(parse_top_level(source));
}

#[test]
pub fn top_level_3() {
    let source = r"
infixl 1 add as +

example = if if a then b else c then if d then e else f else if g then h else i
";

    insta::assert_debug_snapshot!(parse_top_level(source));
}

#[test]
pub fn top_level_4() {
    let source = r"
example = f a b
";

    insta::assert_debug_snapshot!(parse_top_level(source));
}

#[test]
pub fn top_level_5() {
    let source = r"
example = f if a then b else c d

example = f (if a then b else c) d
";

    insta::assert_debug_snapshot!(parse_top_level(source));
}
