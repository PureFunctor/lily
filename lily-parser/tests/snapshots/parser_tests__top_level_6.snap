---
source: lily-parser/tests/parser_tests.rs
expression: parse_top_level(source)
---
Ok(
    Module {
        declarations: [
            Declaration {
                begin: 1,
                end: 118,
                kind: ValueDeclaration(
                    "example",
                    [],
                    Expression {
                        begin: 11,
                        end: 118,
                        kind: DoBlock(
                            [
                                DoStatement {
                                    begin: 16,
                                    end: 41,
                                    kind: LetStatement(
                                        [
                                            Declaration {
                                                begin: 24,
                                                end: 30,
                                                kind: ValueDeclaration(
                                                    "u",
                                                    [],
                                                    Expression {
                                                        begin: 28,
                                                        end: 30,
                                                        kind: Integer(
                                                            "21",
                                                        ),
                                                    },
                                                ),
                                            },
                                            Declaration {
                                                begin: 35,
                                                end: 41,
                                                kind: ValueDeclaration(
                                                    "v",
                                                    [],
                                                    Expression {
                                                        begin: 39,
                                                        end: 41,
                                                        kind: Integer(
                                                            "21",
                                                        ),
                                                    },
                                                ),
                                            },
                                        ],
                                    ),
                                },
                                DoStatement {
                                    begin: 44,
                                    end: 56,
                                    kind: BindExpression(
                                        LesserPattern {
                                            begin: 44,
                                            end: 45,
                                            kind: Variable(
                                                "w",
                                            ),
                                        },
                                        Expression {
                                            begin: 49,
                                            end: 56,
                                            kind: Application(
                                                Expression {
                                                    begin: 49,
                                                    end: 53,
                                                    kind: Variable(
                                                        "pure",
                                                    ),
                                                },
                                                [
                                                    Expression {
                                                        begin: 54,
                                                        end: 56,
                                                        kind: Integer(
                                                            "21",
                                                        ),
                                                    },
                                                ],
                                            ),
                                        },
                                    ),
                                },
                                DoStatement {
                                    begin: 59,
                                    end: 71,
                                    kind: BindExpression(
                                        LesserPattern {
                                            begin: 59,
                                            end: 60,
                                            kind: Variable(
                                                "x",
                                            ),
                                        },
                                        Expression {
                                            begin: 64,
                                            end: 71,
                                            kind: Application(
                                                Expression {
                                                    begin: 64,
                                                    end: 68,
                                                    kind: Variable(
                                                        "pure",
                                                    ),
                                                },
                                                [
                                                    Expression {
                                                        begin: 69,
                                                        end: 71,
                                                        kind: Integer(
                                                            "21",
                                                        ),
                                                    },
                                                ],
                                            ),
                                        },
                                    ),
                                },
                                DoStatement {
                                    begin: 74,
                                    end: 118,
                                    kind: DiscardExpression(
                                        Expression {
                                            begin: 74,
                                            end: 118,
                                            kind: Application(
                                                Expression {
                                                    begin: 74,
                                                    end: 81,
                                                    kind: Variable(
                                                        "attempt",
                                                    ),
                                                },
                                                [
                                                    Expression {
                                                        begin: 82,
                                                        end: 118,
                                                        kind: DoBlock(
                                                            [
                                                                DoStatement {
                                                                    begin: 89,
                                                                    end: 101,
                                                                    kind: BindExpression(
                                                                        LesserPattern {
                                                                            begin: 89,
                                                                            end: 90,
                                                                            kind: Variable(
                                                                                "y",
                                                                            ),
                                                                        },
                                                                        Expression {
                                                                            begin: 94,
                                                                            end: 101,
                                                                            kind: Application(
                                                                                Expression {
                                                                                    begin: 94,
                                                                                    end: 98,
                                                                                    kind: Variable(
                                                                                        "pure",
                                                                                    ),
                                                                                },
                                                                                [
                                                                                    Expression {
                                                                                        begin: 99,
                                                                                        end: 101,
                                                                                        kind: Integer(
                                                                                            "21",
                                                                                        ),
                                                                                    },
                                                                                ],
                                                                            ),
                                                                        },
                                                                    ),
                                                                },
                                                                DoStatement {
                                                                    begin: 106,
                                                                    end: 118,
                                                                    kind: BindExpression(
                                                                        LesserPattern {
                                                                            begin: 106,
                                                                            end: 107,
                                                                            kind: Variable(
                                                                                "z",
                                                                            ),
                                                                        },
                                                                        Expression {
                                                                            begin: 111,
                                                                            end: 118,
                                                                            kind: Application(
                                                                                Expression {
                                                                                    begin: 111,
                                                                                    end: 115,
                                                                                    kind: Variable(
                                                                                        "pure",
                                                                                    ),
                                                                                },
                                                                                [
                                                                                    Expression {
                                                                                        begin: 116,
                                                                                        end: 118,
                                                                                        kind: Integer(
                                                                                            "21",
                                                                                        ),
                                                                                    },
                                                                                ],
                                                                            ),
                                                                        },
                                                                    ),
                                                                },
                                                            ],
                                                        ),
                                                    },
                                                ],
                                            ),
                                        },
                                    ),
                                },
                            ],
                        ),
                    },
                ),
            },
        ],
    },
)
